### PR TITLE
Improve string fstring coverage

### PR DESCRIPTION
**Pull request by dvermd**
The test_fstring.py file comes from CPython

This improve coverage as of #1671.

### CODE REVIEW

**Comment by youknowone** [comment]
@charliermarsh could you review this PR when it is ready?

**Comment by fanninpm** [comment]
There are currently a bunch of unexpectedly passing tests in `test_ast.py` and `test_tokenize.py`.

**Comment by dvermd** [comment]
The change on python.lalrpop might be extended to all binop. What do you think ?

File: Lib/test/test_global.py [hunk 1]
```diff
@@ -25,15 +25,15 @@ def wrong1():
     global a
     global b
 """
-        check_syntax_error(self, prog_text_1, lineno=4, offset=5)
+        check_syntax_error(self, prog_text_1, lineno=4, offset=4)
```
Comment (youknowone) [inline Lib/test/test_global.py:28]: We mostly try not to make changes in cpython-originated code.
Please keep this as original code and add `@unittest.expectedFailure` if you cannot keep this compatibility with the patch.

File: compiler/core/src/location.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ impl Location {
 
     pub fn reset(&mut self) {
         self.row = 1;
-        self.column = 1;
+        self.column = 0;
```
Comment (youknowone) [inline compiler/core/src/location.rs:46]: maybe this is the reason `test_global` is broken.

File: compiler/parser/src/fstring.rs [hunk 3]
```diff
@@ -67,10 +67,18 @@ impl FStringParser {
                         Some('a') => ConversionFlag::Ascii,
                         Some('r') => ConversionFlag::Repr,
                         Some(_) => {
-                            return Err(InvalidConversionFlag);
+                            if expression[1..].trim().is_empty() {
+                                return Err(EmptyExpression);
+                            } else {
+                                return Err(InvalidConversionFlag);
+                            }
```
Comment (youknowone) [inline compiler/parser/src/fstring.rs:74]:
```suggestion
                            return Err(if expression[1..].trim().is_empty() {
                                EmptyExpression
                            } else {
                                InvalidConversionFlag
                            };
```
This is also possible

**Review by youknowone — COMMENTED** [review]
Thank you for contributing. I left a few comments about policy and styles.

@charliermarsh Could you please do the technical review?

File: compiler/core/src/location.rs (hunk 2 above)
Comment (dvermd) [reply compiler/core/src/location.rs:46]: That change make so many other test pass for example in `test_ast` that I don't think the `test_global` error comes from it.
There may have been 2 problems hiding/cancelling each other for `test_global` to be OK.

File: Lib/test/test_global.py (hunk 1 above)
Comment (dvermd) [reply Lib/test/test_global.py]: fixed

**Comment by fanninpm** [comment]
Can you split d34d361 into two separate commits: one commit adding the test file, and one commit adding the RustPython changes?

**Comment by charliermarsh** [comment]
(I promise to review when I can, if others get to it first no prob.)

**Comment by dvermd** [comment]
> Can you split [d34d361](https://github.com/RustPython/RustPython/commit/d34d361cefc643537c9c2e51eccda07a64b9181c) into two separate commits: one commit adding the test file, and one commit adding the RustPython changes?

Done

**Comment by dvermd** [comment]
Looks like some more unexpected success appeared

File: Lib/test/test_fstring.py [hunk 4]
```diff
@@ -0,0 +1,1392 @@
...
+        self.assertFalse(x.called)
+
+        # Actually run the code.
+        exec(c)
+
+        # Make sure x was called.
+        self.assertTrue(x.called)
+
+    # TODO: RUSTPYTHON - binop lineno
+    @unittest.expectedFailure
```
Comment (fanninpm) [inline Lib/test/test_fstring.py:87]: These two lines, and pairs of lines like this, should live in a separate commit.

Comment (dvermd) [reply Lib/test/test_fstring.py:87]: commit split and branch rebased on main

File: Lib/test/test_global.py (hunk 1 above)
Comment (dvermd) [reply Lib/test/test_global.py]: I think this change might be handle the same way this CPython commit handles it [025eb98dc0](https://github.com/python/cpython/commit/025eb98dc0c1dc27404df6c544fc2944e0fa9f3a): make all syntax error 1-based by incrementing location in "constructor" by 1

Comment (youknowone) [reply Lib/test/test_global.py]: Oh, thank you! good to know.

**Comment by youknowone** [comment]
which version of CPython did you use for https://github.com/RustPython/RustPython/pull/4231/commits/d48d998e298a75c0f8ea9d629d0a77835f084219 ?

File: compiler/core/src/location.rs (hunk 2 above)
Comment (charliermarsh) [reply compiler/core/src/location.rs]: I guess that's consistent with the CPython `ast` module, but it seems like a fairly significant change.

File: compiler/parser/src/fstring.rs [hunk 5]
```diff
@@ -245,16 +285,22 @@ impl FStringParser {
...
                     if nested == 0 {
-                        if let Some('{') = chars.peek() {
-                            chars.next();
-                            content.push('{');
-                            continue;
+                        match chars.peek() {
+                            Some('{') => {
+                                chars.next();
+                                content.push('{');
+                                // brace_opened = true;
```
[the remark this answers is not in the document]
Comment (dvermd) [reply compiler/parser/src/fstring.rs]: This is fixed now

**Comment by dvermd** [comment]
> which version of CPython did you use for [d48d998](https://github.com/RustPython/RustPython/commit/d48d998e298a75c0f8ea9d629d0a77835f084219) ?

I took it from [CPython 3.10.8](https://github.com/python/cpython/blob/v3.10.8/Lib/test/test_fstring.py)

**Comment by fanninpm** [comment]
> > which version of CPython did you use for [d48d998](https://github.com/RustPython/RustPython/commit/d48d998e298a75c0f8ea9d629d0a77835f084219) ?
> 
> I took it from [CPython 3.10.8](https://github.com/python/cpython/blob/v3.10.8/Lib/test/test_fstring.py)

I think @youknowone wants you to amend the commit name to mention that version.

**Comment by dvermd** [comment]
> Is it intended that all locations are now 0-indexed rather than 1-indexed?

@charliermarsh It looks like it's how CPython handles it regarding how it improve coverage in test_ast.py. 

Also @youknowone already merged this  part of the PR.

**Comment by dvermd** [comment]
> > > which version of CPython did you use for [d48d998](https://github.com/RustPython/RustPython/commit/d48d998e298a75c0f8ea9d629d0a77835f084219) ?
> > 
> > 
> > I took it from [CPython 3.10.8](https://github.com/python/cpython/blob/v3.10.8/Lib/test/test_fstring.py)
> 
> I think @youknowone wants you to amend the commit name to mention that version.

commit amended with the CPython version

**Review by youknowone — APPROVED** [review]
@dvermd Thank you! This is a big progress!

@charliermarsh @fanninpm Thank you for the reviews!

**Pull request merged**